File: sample-chat/sample-chat/Services/ConnectionManager.m
//
//  ConnectionManager.m
//  Sample-chat
//
//  Created by Andrey Ivanov on 12.12.14.
//  Copyright (c) 2014 QuickBlox Team. All rights reserved.
//

#import "ConnectionManager.h"
#import "QBServiceManager.h"
#import "StorageManager.h"

@interface ConnectionManager()

@property (copy, nonatomic) void(^chatLoginCompletionBlock)(BOOL success, NSString *errorMessage);
@end

@implementation ConnectionManager

+ (instancetype)instance {
	
	static id instance = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		instance = [[self alloc] init];
	});
	
	return instance;
}

- (instancetype)init {
	self = [super init];
	if( self ) {
		
	}
	return self;
}
#pragma mark - Login / Logout

- (void)logInWithUser:(QBUUser *)user
		   completion:(void (^)(BOOL success, NSString *errorMessage))completion {
	
	__weak __typeof(self)weakSelf = self;
	[[QBServiceManager instance].authService logInWithUser:user completion:^(QBResponse *response, QBUUser *userProfile) {
		
		if( response.error != nil ){
			completion(NO, response.error.error.localizedDescription);
			return;
		}
		if( QBServiceManager.instance.currentUser != nil ){
			[QBServiceManager.instance.chatService logoutChat];
			[StorageManager.instance reset];
		}
		
		[QBServiceManager.instance.chatService logIn:^(NSError *error) {
			completion(error == nil, error.localizedDescription);
		}];
		
	}];
}

- (void)logOut {
	
}

#pragma mark - QBChatDelegate

- (void)chatDidNotLogin {
	if (self.chatLoginCompletionBlock) {
		self.chatLoginCompletionBlock(NO, @"");
		self.chatLoginCompletionBlock = nil;
	}
}

- (void)chatDidFailWithError:(NSInteger)code {
	if (self.chatLoginCompletionBlock) {
		self.chatLoginCompletionBlock(NO, @"");
		self.chatLoginCompletionBlock = nil;
	}
}

- (void)chatDidLogin {
	if (self.chatLoginCompletionBlock) {
		self.chatLoginCompletionBlock(YES, @"");
		self.chatLoginCompletionBlock = nil;
	}
}

@end

@implementation QBUUser (ConnectionManager)

- (NSUInteger)index {
	NSUInteger idx = [QBServiceManager.instance.usersService indexOfUser:self];
	return idx;
}

- (UIColor *)color {
	UIColor *color = [QBServiceManager.instance.usersService colorForUser:self];
	return color;
}

@end

File: sample-chat-swift/sample-chat-swift/Classes/Services/ConnectionManager.swift
//
//  ConnectionManager.swift
//  sample-chat-swift
//
//  Created by Anton Sokolchenko on 3/31/15.
//  Copyright (c) 2015 quickblox. All rights reserved.
//

import UIKit

class ConnectionManager: NSObject, QBChatDelegate {
	static let instance = ConnectionManager()

	var timer: NSTimer? // join all rooms, handle internet connection availability
	
	let privacyManager:PrivacyManager = PrivacyManager()
	
	private override init() {
		super.init()
	}
	
	func logInWithUser(user: QBUUser, completion: (success: Bool, errorMessage: String?) -> Void){
        
        ServicesManager.instance.setupChatCacheService(user.login)
        
        ServicesManager.instance.authService.logInWithUser(user, completion: { (response:QBResponse!, user: QBUUser!) -> Void in
            
            if (response.error != nil) {
                
                completion(success: false, errorMessage: response.error.error.localizedDescription)
                return
            }
            
            if ServicesManager.instance.currentUser() != nil {
                
                ServicesManager.instance.chatService.logoutChat()
                StorageManager.instance.reset()
                self.privacyManager.reset()
                
            }
            
            ServicesManager.instance.chatService.logIn({ (error : NSError!) -> Void in
                
                if error == nil {
                    
                    self.privacyManager.retrieveDefaultPrivacyList()
                }
                
                completion(success: error == nil, errorMessage: error?.localizedDescription)
                
            })
            
        })
        
	}
	
	func joinAllRooms() {
        
		if !IJReachability.isConnectedToNetwork() {
			return
		}
        
		if !QBChat.instance().isLoggedIn() {
			return
		}
        
        let dialogs = StorageManager.instance.dialogs
        
		if dialogs.count > 0 {
			let groupDialogs = dialogs.filter({$0.type != .Private})
            
			for roomDialog in groupDialogs {
				if !roomDialog.chatRoom.isJoined {
					roomDialog.chatRoom.joinRoomWithHistoryAttribute(["maxstanzas":0])
				}
			}
		}
		
	}
}
